presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -173,7 +219,11 @@ public Block readPrimitive(ColumnDescriptor columnDescriptor, Type type)
             ParquetColumnChunk columnChunk = new ParquetColumnChunk(descriptor, buffer, 0);
             columnReader.setPageReader(columnChunk.readAllPages());
         }
-        return columnReader.readPrimitive(type);
+        Block block = columnReader.readPrimitive(type, offsets);
+        if (positions.isEmpty()) {
+            return block;
+        }
+        return block.copyPositions(positions);
```

Comment:
I'm having a hard time understand this code.  I don't really understand how `offsets` and `positions` are used, and I really don't understand why you would have copyPositions, which is super expensive, here.